drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1901 @@
+		/* Cancel EP on-going transaction */
+		USBD_STOP_TRANSACTION(EP0 + i);
+
+		/* Reset EP to unstalled */
+		USBD_CLR_EP_STALL(EP0 + i);
+
+		/* Reset EP data toggle bit to 0 */
+		USBD_SET_DATA0(EP0 + i);
+
+		/* Except EP0/EP1 kept resident for CTRL OUT/IN, disable all other EPs */
+		if (i >= 2) {
+			USBD_CONFIG_EP(EP0 + i, 0);
+		}
+	}
+
+	/* Reset USB device address to 0 */
+	m_numaker_usbd_reset_addr(dev);
+}
+
+/* Signal remote wakeup */
+static void m_numaker_usbd_remote_wakeup(struct usb_dc_numaker_device *dev)
+{
+	const struct usb_dc_numaker_config *config = dev->config;
+
+	/* Enable back USB/PHY first, then generate 'K' >= 1 ms (USB spec) */
+	USBD_ENABLE_USB();
+
+	config->usbd_base->ATTR |= USBD_ATTR_RWAKEUP_Msk;
+	k_busy_wait(1000);
+	config->usbd_base->ATTR ^= USBD_ATTR_RWAKEUP_Msk;
+}
+
+/* Copy to user buffer from Setup FIFO */
+static void m_numaker_usbd_setup_fifo_copy_to_user(struct usb_dc_numaker_device *dev,
+						   uint8_t *usrbuf)
+{
+	const struct usb_dc_numaker_config *config = dev->config;
+
+	uint32_t dmabuf_addr =
+		USBD_BUF_BASE + (config->usbd_base->STBUFSEG & USBD_STBUFSEG_STBUFSEG_Msk);
+	USBD_MemCopy(usrbuf, (uint8_t *)dmabuf_addr, 8ul);
+}
+
+/* Reset EP FIFO
+ *
+ * NOTE: EP FIFO is based on EP DMA buffer, which may not be configured yet.
+ */
+static void numaker_usbd_ep_fifo_reset(struct numaker_usbd_ep *ep_cur)
+{
+	if (USB_EP_DIR_IS_OUT(ep_cur->ep_addr)) {
+		/* Read FIFO */
+		if (ep_cur->dmabuf_valid) {
+			ep_cur->read_fifo_pos = ep_cur->dmabuf_base;
+			ep_cur->read_fifo_used = 0;
+		}
+	} else {
+		/* Write FIFO */
+		if (ep_cur->dmabuf_valid) {
+			ep_cur->write_fifo_pos = ep_cur->dmabuf_base;
+			ep_cur->write_fifo_free = numaker_usbd_ep_fifo_max(ep_cur);
+		}
+	}
+}
+
+/* Copy data to user buffer from EP FIFO
+ *
+ * size_p holds size to copy/copied on input/output
+ */
+static int numaker_usbd_ep_fifo_copy_to_user(struct numaker_usbd_ep *ep_cur, uint8_t *usrbuf,
+					     uint32_t *size_p)
+{
+	__ASSERT_NO_MSG(size_p);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+
+	/* Clamp to read FIFO used count */
+	*size_p = Z_MIN(*size_p, numaker_usbd_ep_fifo_used(ep_cur));
+
+	uint32_t dmabuf_addr = USBD_BUF_BASE + ep_cur->read_fifo_pos;
+
+	USBD_MemCopy(usrbuf, (uint8_t *)dmabuf_addr, *size_p);
+
+	/* Advance read FIFO */
+	ep_cur->read_fifo_pos += *size_p;
+	ep_cur->read_fifo_used -= *size_p;
+	if (ep_cur->read_fifo_used == 0) {
+		ep_cur->read_fifo_pos = ep_cur->dmabuf_base;
+	}
+
+	return 0;
+}
+
+/* Copy data from user buffer to EP FIFO
+ *
+ * size_p holds size to copy/copied on input/output
+ */
+static int numaker_usbd_ep_fifo_copy_from_user(struct numaker_usbd_ep *ep_cur,
+					       const uint8_t *usrbuf, uint32_t *size_p)
+{
+	__ASSERT_NO_MSG(size_p);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	__ASSERT_NO_MSG(ep_cur->ep_mps <= ep_cur->dmabuf_size);
+
+	*size_p = Z_MIN(*size_p,
+			numaker_usbd_ep_fifo_max(ep_cur) - numaker_usbd_ep_fifo_used(ep_cur));
```

Comment:
`Z_MIN` is GCC only, you  may want to use `MIN` unless you really want to avoid double evaluation, though in that case you could also use a temoprary variable

Reply:
changed to `MIN`